scripts/docker/docker.lib.sh

```diff
@@ -0,0 +1,256 @@
+    else
+      # Pull the latest image
+      docker pull \
+        --platform linux/amd64 \
+        "${name}:latest" \
+      > /dev/null 2>&1 || true
+    fi
+  fi
+
+  echo "${name}:${version}"
+}
+
+# ==============================================================================
+# "Private" functions.
+
+# Create effective Dockerfile.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function _create-effective-dockerfile() {
+
+  local dir=${dir:-$PWD}
+  cp "${dir}/Dockerfile" "${dir}/Dockerfile.effective"
+  _replace-image-latest-by-specific-version
+  _append-metadata
+}
+
+# Replace image:latest by a specific version.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function _replace-image-latest-by-specific-version() {
+
+  local dir=${dir:-$PWD}
+  versions_file=$(git rev-parse --show-toplevel)/.tool-versions
+  if [ -f "$versions_file" ]; then
+    # First, list the entries specific for Docker to take precedence, then the rest
+    content=$(grep " docker/" "$versions_file"; grep -v " docker/" "$versions_file")
+    echo "$content" | while IFS= read -r line; do
+      [ -z "$line" ] && continue
+      line=$(echo "$line" | sed "s/^#\s*//; s/\s*#.*$//" | sed "s;docker/;;")
+      name=$(echo "$line" | awk '{print $1}')
+      version=$(echo "$line" | awk '{print $2}')
+      sed -i "s;FROM ${name}:latest;FROM ${name}:${version};g" "${dir}/Dockerfile.effective"
+    done
+  fi
+}
+
+# Append metadata to the end of Dockerfile.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function _append-metadata() {
+
+  local dir=${dir:-$PWD}
+  cat \
+    "$dir/Dockerfile.effective" \
+    "$(git rev-parse --show-toplevel)/scripts/docker/Dockerfile.metadata" \
+  > "$dir/Dockerfile.effective.tmp"
+  mv "$dir/Dockerfile.effective.tmp" "$dir/Dockerfile.effective"
+}
+
+# Create effective version from the VERSION file.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function _create-effective-version() {
+
+  local dir=${dir:-$PWD}
+  build_datetime=${BUILD_DATETIME:-$(date -u +'%Y-%m-%dT%H:%M:%S%z')}
+  if [ -f "$dir/VERSION" ]; then
+    # shellcheck disable=SC2002
+    cat "$dir/VERSION" | \
+      sed "s/yyyy/$(date --date="${build_datetime}" -u +"%Y")/g" | \
+      sed "s/mm/$(date --date="${build_datetime}" -u +"%m")/g" | \
+      sed "s/dd/$(date --date="${build_datetime}" -u +"%d")/g" | \
+      sed "s/HH/$(date --date="${build_datetime}" -u +"%H")/g" | \
+      sed "s/MM/$(date --date="${build_datetime}" -u +"%M")/g" | \
+      sed "s/SS/$(date --date="${build_datetime}" -u +"%S")/g" | \
+      sed "s/hash/$(git rev-parse --short HEAD)/g" \
+    > "$dir/.version"
```

Comment:
Thanks @regularfry. That could be an option.

It is worth mentioning that we would like to use the same versioning method for the repository code itself and other artefacts.

The version pattern is stored in an artefact `VERSION` file. This file is processed by the automation script and converted to an effective file saved as `.version` in the same directory. Then the relevant "tag" or "tags" are created based on its content. However, these tags are not a complete set of tags that we would expect to assign to an artefact or a Git commit, e.g. the artefact promotion process to higher instance environments would create more tags, but not versions. If we use `.tags` or `.tag` instead of the `.version` file, this may open the process to misuse. Tagging is a separate step and I would like to keep a distinction between the two.

Most projects will use a single version pattern, e.g. `yyyymmdd` (I think we can agree that semantic versioning doesn't work here and manual version changes are no longer feasible), with exceptions for those that deploy multiple times a day and whose deployment cycle doesn't directly match the release cycle. Therefore, despite the functionality supporting multiple versions being assigned, I think the concept will be simpler to grasp if we keep the name of the versioning file singular, i.e. `VERSION` and `.version`.

I have [updated the documentation](https://github.com/nhs-england-tools/repository-template/blob/c32f7042ecd24a46ab3c7fe3b65e1804881357a4/docs/developer-guides/Scripting_Docker.md#docker) to explain how it works. Do we think this helps?

Reply:
Yes, I think that's fine, although it would be very nice if we could provide a nudge towards multiple releases per day with less-than-`dd` resolution in the tag.  That's not vital though.